### PR TITLE
Nvidia dali

### PR DESCRIPTION
**Pull request by Shrey-Viradiya**
Integration for NVIDIA DALI completed. Drastically improved performance. From nearly 130 sec per epoch to 8 sec per epoch.

### CODE REVIEW

File: NvidiaDali.py [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+        self.RRC = ops.RandomResizedCrop(size = output_size,
+            random_area = [0.4, 1.0],
+            random_aspect_ratio = [0.5, 1.5],
+            device="gpu"
+        )
+        self.cmn = ops.CropMirrorNormalize(
+        device="gpu",
+        dtype=types.FLOAT,
+        mean=[0.485, 0.456, 0.406], std=[0.229, 0.224, 0.225],
+        output_layout="HWC")
```
Comment (tirthasheshpatel) [inline NvidiaDali.py:20]:
```suggestion
        self.cmn = ops.CropMirrorNormalize(
                       device="gpu",
                       dtype=types.FLOAT,
                       mean=[0.485, 0.456, 0.406],
                       std=[0.229, 0.224, 0.225],
                       output_layout="HWC"
                   )
```
Small indentation nitpick

File: NvidiaDali.py [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+class HybridPipelineTest(Pipeline):
+    def __init__(self, batch_size, output_size, num_threads, device_id, images_directory):
+        super(HybridPipelineTest, self).__init__(batch_size, num_threads, device_id, seed = 12)
+        self.input = ops.FileReader(file_root = images_directory, random_shuffle = True, initial_fill = 21)
+        self.decode = ops.ImageDecoder(device = "mixed", output_type = types.RGB)
+        self.cmn = ops.CropMirrorNormalize(
+        device="gpu",
+        dtype=types.FLOAT,
+        mean=[0.485, 0.456, 0.406], std=[0.229, 0.224, 0.225],
+        output_layout="HWC")
```
Comment (tirthasheshpatel) [inline NvidiaDali.py:45]:
```suggestion
        self.cmn = ops.CropMirrorNormalize(
                       device="gpu",
                       dtype=types.FLOAT,
                       mean=[0.485, 0.456, 0.406],
                       std=[0.229, 0.224, 0.225],
                       output_layout="HWC"
                   )
```
Small indentation nitpick

File: README.md [hunk 3]
```diff
@@ -30,14 +30,18 @@ This repository contains code for pneumonia detection using X-ray images of the
         - GoogleNet
         - Inception
     - If using colab for training, mount the drive and use --colab to save the files in the drive
+    - For faster training and Image augmentation, if on Linux, NVIDIA DALI can be used
+        - Install NVIDIA DALI
```
Comment (tirthasheshpatel) [inline README.md:34]:
```suggestion
        - Install NVIDIA DALI. Please see the [installation page](https://docs.nvidia.com/deeplearning/dali/user-guide/docs/installation.html) for details.
```

File: README.md [hunk 4]
```diff
@@ -30,14 +30,18 @@ This repository contains code for pneumonia detection using X-ray images of the
...
+        - Note that NVIDIA DALI is only available for Linux
     - To change optimizers, use one of the following
         - Adam
         - SGD
         - RMSprop
         - Adagrad
         - Adadelta
 - To test new dataset, run `test.py` with directory path and base model with option given above. Make sure that model is trained on those models first. 
-    - Example, `python3 test.py "./data/Corona_Classification_data/test/" Inception`
+    - Example, `python3 test.py "./data/Corona_Classification_data/test/" Inception True True`
```
Comment (tirthasheshpatel) [inline README.md:44]: Can you also explain what those two `True` are and what they do?

Oh, never mind: That ar explained in the usage below!

**Review by tirthasheshpatel — APPROVED** [review]
Just a couple of non-blocking nitpicks. Otherwise looks good.

**Comment by tirthasheshpatel** [comment]
I have applied all the suggestions. So, merging. Thanks, @Shrey-Viradiya!

**Comment by Shrey-Viradiya** [comment]
> I have applied all the suggestions. So, merging. Thanks, @Shrey-Viradiya!

Thanks @tirthasheshpatel

**Pull request merged**